source/validate_instruction.cpp

```diff
@@ -145,14 +146,51 @@ spv_result_t LimitCheckIdBound(ValidationState_t& _,
 // Checks that the number of OpTypeStruct members is within the limit.
 spv_result_t LimitCheckStruct(ValidationState_t& _,
                               const spv_parsed_instruction_t* inst) {
+  if (SpvOpTypeStruct != inst->opcode) {
+    return SPV_SUCCESS;
+  }
+
   // Number of members is the number of operands of the instruction minus 1.
   // One operand is the result ID.
   const uint16_t limit = 0x3fff;
-  if (SpvOpTypeStruct == inst->opcode && inst->num_operands - 1 > limit) {
+  if (inst->num_operands - 1 > limit) {
     return _.diag(SPV_ERROR_INVALID_BINARY)
            << "Number of OpTypeStruct members (" << inst->num_operands - 1
            << ") has exceeded the limit (" << limit << ").";
   }
+
+  // Section 2.17 of SPIRV Spec specifies that the "Structure Nesting Depth"
+  // must be less than or equal to 255.
+  // This is interpreted as structures including other structures as members.
+  // The code does not follow pointers or look into arrays to see if we reach a
+  // structure downstream.
+  //
+  // This lambda takes a valid struct ID and finds its nesting depth.
+  // The nesting depth of a struct is 1+(largest depth of any member).
+  // Scalars are at depth 0.
+  std::function<int(uint32_t)> find_struct_depth_recursive = [&](uint32_t id) {
```

Comment:
This algorithm suffers from an exponential runtime vulnerability.

Example, 
```struct S0 { int; int };
struct S1 { S0; S0; };
struct S2 { S2; S2; };
...
struct S40 { S39; S39 };
```
To compute the nesting depth of S40, this algorithm will take  about 2**40 steps.

Better to use memoization.  The simple way to do it is to compute and store depths as you scan the module.  This works because the members of a struct are always defined before the struct itself.
(And an OpTypeForwardPointer is just a pointer, which is a leaf anyway.)
This implies an extra member for the validation state.  Probably a std::unordered_map from result ID of struct definitions to depth.  When an ID is not in the map then use the value 0 in the reduction.